### PR TITLE
Don't stop Adaptive on error

### LINKED ISSUE #8808: Adaptive stops whenever it encounters an error
`Adaptive.adapt()` stops the callback whenever it encounters an error while _not scaling down_. Instead, it should be able to recover from spurious errors, but stop when the cluster has been shut down or is unresponsive for a longer period of time.

https://github.com/dask/distributed/blob/564f28b5637b56fd2e6870c4489d299f8984576a/distributed/deploy/adaptive_core.py#L274-L280

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #8808

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by jacobtomlinson — APPROVED** [review]
This seems fine to me. It's not clear from reading #2904 why that stop was introduced in the first place.

**Comment by hendrikmakait** [comment]
> This seems fine to me. It's not clear from reading https://github.com/dask/distributed/pull/2904 why that stop was introduced in the first place.

My gut feeling is that it was introduced to cover the case where the cluster closes, and the scheduler becomes unavailable. I think this should be covered by the `Cluster` object (and is indeed handled by `Cluster`). The open question is how to enforce this from subclasses or duck-typed clusters. (This is why I moved this to draft.)

**Comment by jacobtomlinson** [comment]
Yeah I feel like the cluster object should stop adapting if the scheduler connection closes. I think we want to implement a protocol for cluster objects that sets out expectations and formalises the API a little. But for now I think implementing it in `SpecCluster` should be enough?

**Comment by hendrikmakait** [comment]
@jacobtomlinson: Conceptually, I agree with your approach, but for now I'd like to go the more conversative route of having Adaptive itself check whether its cluster is still running. This has caused the PR to blow up a bit because some functionality and tests needed to move between `adaptive` and `adaptive_core`.

**Review by jacobtomlinson — COMMENTED** [review]
This looks great, thanks for taking the time here Hendrik.

Have you tried these changes out with any of the projects that use adaptive scaling? AFAIK that's only `dask-jobqueue` and `dask-cloudprovider` today.

File: distributed/deploy/adaptive.py [hunk 1]
```diff
@@ -81,16 +100,21 @@ class Adaptive(AdaptiveCore):
...
-        minimum=None,
-        maximum=None,
-        wait_count=None,
-        target_duration=None,
-        worker_key=None,
-        **kwargs,
+        cluster: Cluster,
+        interval: str | float | timedelta | None = None,
+        minimum: int | None = None,
+        maximum: float | None = None,
```
Comment (ntabris) [inline distributed/deploy/adaptive.py:113]: Hm, this is a little confusing given that we actually either want `int` or `math.inf`, per https://github.com/dask/distributed/blob/8bafad5450592d9a647080589af1ca3072cc5341/distributed/deploy/adaptive_core.py#L112-L113

Comment (hendrikmakait) [reply distributed/deploy/adaptive.py:113]: I don't have strong feelings here, this just follow the numeric tower in [PEP 484](https://peps.python.org/pep-0484/#the-numeric-tower), i.e., `int | float` can be simplified to `float`. Unfortunately there isn't a canonical `INF` literal that I'm aware of, instead we have `math.inf`, `float("inf")`, `np.inf`, and more. Happy to change it back if people find it confusing.

Comment (jacobtomlinson) [reply distributed/deploy/adaptive.py:113]: Intuitively I would say that `None` is equivalent to `math.inf` in this case. I expect most users would assume that setting the value to `None` would not set an upper bound. They might even assume that setting it to `0` does the same.

Should we attempt to define an `INF` literal in `dask.typing` that we can use here?

Comment (hendrikmakait) [reply distributed/deploy/adaptive.py:113]: Sounds reasonable but let's do that in a different PR. I'll roll my changes to the type back for now.

Comment (hendrikmakait) [reply distributed/deploy/adaptive.py]: Note that `None` is not equivalent to `math.inf` in this case. Instead, we read the config value if `None` is provided.

**Comment by hendrikmakait** [comment]
> Have you tried these changes out with any of the projects that use adaptive scaling? AFAIK that's only `dask-jobqueue` and `dask-cloudprovider` today.

I've executed both test suites locally once without failures. There were quite a few skipped tests, so YMMV.

**Review by jacobtomlinson — APPROVED** [review]

**Comment by jacobtomlinson** [comment]
🎉

**Pull request merged**